File: Python/rate_limits.py
from asyncio import ALL_COMPLETED, Future
import time
from amadeus import Client, ResponseError
from dotenv import load_dotenv
import datetime
import concurrent.futures
from limiter import RateLimitException, limiter

#Read environment variables from .env file
load_dotenv()

amadeus = Client()

departure_date = datetime.datetime.now()
return_date = (departure_date + datetime.timedelta(days=10)).strftime('%Y-%m-%d')
departure_date = departure_date.strftime('%Y-%m-%d')

nb_requests = 10
successful = 0
failed = 0
search_params = {
  'originLocationCode': 'MAD',
  'destinationLocationCode': 'LHR',
  'departureDate': departure_date,
  'returnDate': return_date,
  'adults': 1,
  'travelClass': 'ECONOMY'
}
start = time.time()


# Using ratelimit package we can decorate a function with @sleep_and_retry
# so the thread executing the function will sleep after the number of calls
# defined in the @limits decorator is reached within the period also defined
# in the @limits decorator
@limiter(calls=50, refresh=5, time_between=1)
def flights_offers_search(**kwargs):
  elapsed = time.time() - start
  print(f'Request sent after {elapsed:.3f}s', flush=True)
  return amadeus.shopping.flight_offers_search.get(**kwargs)


print('Sending requests...')
results = []

for i in range(0,15):
  results.append(flights_offers_search(**search_params))


concurrent.futures.wait(results, timeout=None, return_when=ALL_COMPLETED)

for res in results:
  print(res.Response)


# Print all successes

# for result in results:
#   print(results[0].data[0]['price']['grandTotal'])

File: Node/rate_limits.js
require('dotenv').config();
const Bottleneck = require('bottleneck');
const Amadeus = require('amadeus');
const dateFns = require('date-fns')

//Initialize Amadeus client
const amadeus = new Amadeus({
  clientId: process.env.AMADEUS_CLIENT_ID,
  clientSecret: process.env.AMADEUS_CLIENT_SECRET,
});


//Config for test environment
//Rate limits in test environment are 1 request each 100ms and 10 requests per second
//Check the README for configuration information
const limiter = new Bottleneck({
  minTime: 400,
  reservoir: 10000,
  reservoirRefreshInterval: 30 * 24 * 60 * 60 * 1000,
  reservoirRefreshAmount: 10000
});

//Send 50 requests at once using bottleneck
const nbRequests = 200;
const departureDate = dateFns.format(dateFns.addDays(new Date(), 1), 'yyyy-MM-dd');
const returnDate = dateFns.format(dateFns.addDays(new Date(), 20), 'yyyy-MM-dd');
let successful = 0;
let failed = 0;
let start = Date.now();
console.log("Sending requests ...");
for (let i = 0; i < nbRequests; i++) {


  limiter.schedule(() => {
    let sent = (Date.now() - start) / 1000;
    console.log(`Request ${i + 1} sent after ${sent}s`)

    amadeus.shopping.flightOffersSearch.get({
      originLocationCode: 'MAD',
      destinationLocationCode: 'LHR',
      departureDate: departureDate,
      returnDate: returnDate,
      adults: 1,
      travelClass: 'ECONOMY'
    })
      .then(() => successful += 1)
      .catch((err) => {
        if (err.description[0].status === 429) failed += 1;
      }).finally(() => {
        if (successful + failed === nbRequests) {
          console.log(`${successful} successful requests and ${failed} fails`);
        }
      });
  });
}

limiter.stop({ dropWaitingJobs: false });



